File: core/src/main/java/net/cjsah/bot/util/RequestUtil.java
package net.cjsah.bot.util;

import cn.hutool.http.HttpRequest;
import cn.hutool.http.HttpResponse;
import com.mojang.datafixers.util.Either;
import com.mojang.serialization.Codec;

public class RequestUtil {

    public static <T> Either<T, Error> get(String url, Codec<T> codec) {
        return request(HttpRequest.get(url), codec);
    }

    public static <T> Either<T, Error> post(String url, String body, Codec<T> codec) {
        HttpRequest request = HttpRequest.post(url)
            .header("Content-Type", "application/json")
            .body(body);
        return request(request, codec);
    }

    public static <T> Either<T, Error> request(HttpRequest request, Codec<T> codec) {
        try (HttpResponse response = request.execute()) {
            if (response.getStatus() != 200) {
                return Either.right(new Error(response.getStatus(), response.body()));
            }
            String body = response.body();
            Either<T, String> decode = CodecUtil.decode(codec, body);
            if (decode.isLeft()) {
                T value = decode.left().orElseThrow();
                return Either.left(value);
            }
            String error = decode.right().orElseThrow();
            return Either.right(new Error(1500, error));
        }
    }

    public record Error(int code, String message) {
    }
}

File: core/src/main/kotlin/net/cjsah/bot/util/RequestUtil.kt
package net.cjsah.bot.util

import com.alibaba.fastjson2.JSONObject
import io.ktor.client.call.body
import io.ktor.client.request.request
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.HttpMethod
import io.ktor.http.contentType
import net.cjsah.bot.client

object RequestUtil {
    private const val baseUrl = ""

    suspend fun get(url: String): JSONObject {
        return this.resolve(url, HttpMethod.Get)
    }

    suspend fun post(url: String, data: JSONObject): JSONObject {
        return this.resolve(url, HttpMethod.Get, data)
    }

    suspend fun resolve(url: String, requestMethod: HttpMethod, body: JSONObject? = null): JSONObject {
        println(JsonUtil.serialize(body))
        val res = client.request(baseUrl + url) {
            method = requestMethod
            contentType(ContentType.Application.Json)
            if (body != null) {
                setBody(JsonUtil.serialize(body))
            }
        }

        val body = res.body<String>()
        return JsonUtil.deserialize(body)
    }
}